### DISCUSSION TITLE
Managed plot slow down.

### DISCUSSION THREAD

**Discussion by JaminMartin** [pymeasure/pymeasure#405] [Q&A]
Hi, sorry for two questions in one day. 

When using the managed plot (something required to implement this in our lab instead of labview) 

I have noticed a sudden slowdown in the plotting when the data sets get above ~1000 points. Is this due to the cursor being active? I have noticed similar behavior using plotly with large data sets. 

If there is a way to have the plot be a bare-bones x,y plot (turning a lot of these graphics processes off) I would love to know, looking at the pyqtgraph documentation did not seem to help with how it is implemented in pymeasure. 

I am currently running win10, python 3.7.
using this code.     
```
 
import sys
sys.modules['cloudpickle'] = None

import logging
log = logging.getLogger(__name__)
log.addHandler(logging.NullHandler())

import sys
import tempfile
import random
from time import sleep
from pymeasure.log import console_log
from pymeasure.display.Qt import QtGui
from pymeasure.display.windows import ManagedWindow
from pymeasure.experiment import Procedure, Results
from pymeasure.experiment import IntegerParameter, FloatParameter, Parameter

class RandomProcedure(Procedure):

    start = IntegerParameter('start')
    stop = IntegerParameter('stop')
    step = IntegerParameter('step')
    delay = FloatParameter('Delay Time', units='s', default=5)
    seed = Parameter('Random Seed', default='12345')

    DATA_COLUMNS = ['Wavelength', 'Intensity']

    def startup(self):
        log.info("Setting the seed of the random number generator")
        random.seed(self.seed)

    def execute(self):
        log.info("Starting the loop of %d iterations" % self.start)
        for i in range(self.start,self.stop,self.step):
            data = {
                'Wavelength': i,
                'Intensity': random.random()
            }
            self.emit('results', data)
            log.debug("Emitting results: %s" % data)
            sleep(self.delay)
            if self.should_stop():
                log.warning("Caught the stop flag in the procedure")
                break


class MainWindow(ManagedWindow):

    def __init__(self):
        super(MainWindow, self).__init__(
            procedure_class=RandomProcedure,
            inputs=['start','stop','step', 'delay', 'seed'],
            displays=['start','stop','step', 'delay', 'seed'],
            x_axis='Wavelength',
            y_axis='Intensity'
        )
        self.setWindowTitle('GUI Example')

    def queue(self):
        filename = tempfile.mktemp()

        procedure = self.make_procedure()
        results = Results(procedure, filename)
        experiment = self.new_experiment(results)

        self.manager.queue(experiment)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)
    
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())    
```

When the range is set such as 0:1000:1 the plotting becomes quite slow. 

It is also worth noting my progress bars in queue dont update. 

Cheers.

**Comment by jmittelstaedt** [comment]
I can't speak to the plotting slowdown too much, but the reason the progress bars aren't updating is that you need to tell pymeasure what the progress of your `Procedure` is in the `execute` method. You do that by including a line somewhere in the main loop like `self.emit("progress", i/max_iterations)`, I usually cast `i/max_iterations` to an `int` but I'm not sure if that's actually necessary.

**Comment by bilderbuchi** [comment]
This issue has more of a discussion/usage question character, so I will move it to our new discussion forum, which is a better fit than the issue tracker.

**Comment by maederan201** [comment]
I'm having the same or similar problems in my measurements in the lab. This is a nice MWE I guess. In your example, does the entire ManagedWindow also slow down when you choose ranges on the order of (1,10000,1), or when you repeatedly do ranges of (1,1000,1)?

In particular I start noticing it when the plots seem to "load" slowly...

Does anyone have an improvement for this issue?

**Reply by CasperSchippers** [reply]
I also sometimes have this issue, but also noticed that when the graph becomes slow, the measurement also can become slower.
Just now, i came across this discussion: https://stackoverflow.com/questions/46206296/performance-issues-with-pyqtgraph. It is suggested that changing the line-width to 1 (rather than 2 as is used in pymeasure at this moment) the graph can significantly speed up; I check this in the example by @Sacroth, and it indeed seems to work for me: the measurement doesn't slow down anymore and the graph (most notably seen from the crosshair) responds way more smoothly.

**Reply by maederan201** [reply]
Interesting stuff, not at all what I suspected :D

What do you mean by "measurements can become slower"? I usually notice it if I use the sequencer to queue a few measurements at once. Then a single measurement will not be completed until the data is plotted, however if I hide all the measurements (i.e. no data plotting) the sequence runs without slowdown. Or at least seemingly, I haven't timed it or anything.

I will try this linewidth change on my procedures. Would it be something to change permanently in pymeasure or should it happen on user level as needed until the corresponding issue in pyqtgraph is resolved (https://github.com/pyqtgraph/pyqtgraph/issues/533)?

**Reply by CasperSchippers** [reply]
What I meant by "measurements can become slower" is that if you measure the time between data-points, you can see that it increases when more data-point have to be plotted (especially clear when you bring down the delay between data-points, though it also seems depends on the distribution of the data-points). Indeed, if you hide all the measurements, the slow-down disappears and the measurements run just as you would want them to run.

Whether to change it in pymeasure (or at least until the issue in pyqtgraph is fixed) is something to consider. Maybe it would be worth to make it a optional argument? What do you think @bilderbuchi? Should we change this in pymeasure (for the time being, although it seems to me that it will take some time before the issue is resolved in pyqtgraph)?

**Reply by bilderbuchi** [reply]
I think if the pyqtgraph lag is so strong to impact measurements  we should change the linewidth in pymeasure, and leave a comment pointing to the pyqtgraph issue. A fix [seems to be approaching](https://github.com/pyqtgraph/pyqtgraph/pull/2011), but I don't know when that will land in a released pyqtgraph version - at that point we can undo our change and bump the pyqtgraph dependency.
